cmd/juju/charmhub/download.go

```diff
@@ -0,0 +1,221 @@
+}
+
+func (c *downloadCommand) validateCharmOrBundle(charmOrBundle string) error {
+	curl, err := charm.ParseURL(charmOrBundle)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if !charm.CharmHub.Matches(curl.Schema) {
+		return errors.Errorf("%q is not a Charm Hub charm", charmOrBundle)
+	}
+	return nil
+}
+
+// Run is the business logic of the download command.  It implements the meaty
+// part of the cmd.Command interface.
+func (c *downloadCommand) Run(cmdContext *cmd.Context) error {
+	var charmHubURL string
+	if c.charmHubURL != "" {
+		charmHubURL = c.charmHubURL
+	} else {
+		config, err := c.getModelConfig()
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		charmHubURL, _ = config.CharmHubURL()
+	}
+
+	config, err := charmhub.CharmHubConfigFromURL(charmHubURL, downloadLogger{
+		Context: cmdContext,
+	})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	var fileSystem charmhub.FileSystem
+	if c.archivePath != "" {
+		fileSystem = charmhub.DefaultFileSystem()
+	} else {
+		fileSystem = stdoutFileSystem{}
+	}
+
+	client, err := c.getCharmHubClient(config, fileSystem)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	info, err := client.Info(ctx, c.charmOrBundle)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// TODO (stickupkid): All the user to specify the location of the download
+	// path.
+	resourceURL, err := url.Parse(info.DefaultRelease.Revision.Download.URL)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return client.Download(ctx, resourceURL, c.archivePath)
+}
+
+// getAPI returns the API that supplies methods
+// required to execute this command.
+func (c *downloadCommand) getAPI() (ModelConfigGetter, error) {
+	if c.modelConfigAPI != nil {
+		// This is for testing purposes, for testing, both values
+		// should be set.
+		return c.modelConfigAPI, nil
+	}
+	api, err := c.NewAPIRoot()
+	if err != nil {
+		return nil, errors.Annotate(err, "opening API connection")
+	}
+	return modelconfig.NewClient(api), nil
+}
+
+func (c *downloadCommand) getModelConfig() (*config.Config, error) {
+	api, err := c.getAPI()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	attrs, err := api.ModelGet()
+	if err != nil {
+		return nil, errors.Wrap(err, errors.New("cannot fetch model settings"))
+	}
+
+	return config.New(config.NoDefaults, attrs)
+}
+
+func (c *downloadCommand) getCharmHubClient(config charmhub.Config, fileSystem charmhub.FileSystem) (CharmHubClient, error) {
+	if c.charmHubClient != nil {
+		return c.charmHubClient, nil
+	}
+	return charmhub.NewClientWithFileSystem(config, fileSystem)
+}
+
+// CharmHubClient defines a charmhub client, used for querying the charmhub
+// store.
+type CharmHubClient interface {
+	// Info returns charm info on the provided charm name from CharmHub API.
+	Info(context.Context, string) (transport.InfoResponse, error)
+
+	// Download defines a client for downloading charms directly.
```

Comment:
This description has me thinking you'd have to call Download().SomeThingElse, as Download ives you a client?   not a zip file to the location.